### PR TITLE
feature: Materialization of the "main" keypair concept

### PR DESCRIPTION
**Pull request by fregataa**
- Implement `main_keypair` which is unique for every user. the column name is `users.main_access_key`
- Superadmins may change a user's main keypair. But a keypair cannot be deleted while it is a main keypair.
- Update related the CLI and client SDK functions.

**Checklist:** (if applicable)

- [x] Milestone metadata specifying the target backport version
- [ ] API server-client counterparts (e.g., manager API -> client SDK)

### CODE REVIEW

File: src/ai/backend/client/cli/admin/user.py [hunk 1]
```diff
@@ -348,6 +350,12 @@ def add(
         "Note that this feature does not automatically install sudo for the session."
     ),
 )
+@click.option(
+    "--primary-access-key",
```
Comment (lizable) [inline src/ai/backend/client/cli/admin/user.py]: I think `--primary-access-key` flag need to be replaced to `--main-access-key`.

File: src/ai/backend/client/cli/admin/user.py [hunk 2]
```diff
@@ -348,6 +350,12 @@ def add(
...
+    type=OptionalType(str),
+    default=undefined,
+    help="Set primary access key which works as default.",
```
Comment (lizable) [inline src/ai/backend/client/cli/admin/user.py]: Same thing here. "Set main access key ~" would be more natural

File: src/ai/backend/manager/models/user.py [hunk 3]
```diff
@@ -709,6 +747,30 @@ async def _pre_func(conn: SAConnection) -> None:
...
+                    .where(KeyPairRow.access_key == main_access_key)
+                    .options(
+                        noload("*"),
+                        joinedload(KeyPairRow.user_row).options(load_only(UserRow.email)),
+                    )
+                )
+                keypair_row: KeyPairRow | None = (await db_session.scalars(keypair_query)).first()
+                if keypair_row is None:
+                    raise RuntimeError(
+                        "Cannot set non-existing access key as the primary access key."
```
Comment (lizable) [inline src/ai/backend/manager/models/user.py]: I think the word represent "the main keypair" should be unified. in error messages too.

**Review by lizable — CHANGES_REQUESTED** [review]
I just found there are a few minor things about "main" and "primary" thing to be fixed.
Let's use the word "main" instead of "primary", especially the comments I mentioned below.

File: src/ai/backend/manager/registry.py [hunk 4]
```diff
@@ -1762,6 +1762,29 @@ async def create_cluster_ssh_keypair(self) -> ClusterSSHKeyPair:
...
         }
 
+    async def get_user_occupancy(self, user_id, *, db_sess=None):
+        known_slot_types = await self.shared_config.get_resource_slots()
+
+        async def _query() -> ResourceSlot:
+            async with reenter_txn_session(self.db, db_sess) as _sess:
+                subq = sa.select(KeyPairRow.access_key).where(KeyPairRow.user == user_id)
+                query = sa.select(KernelRow.occupied_slots).where(
+                    (KernelRow.access_key.in_(subq))
```
Comment (achimnol) [inline src/ai/backend/manager/registry.py]: I think we could simply filter `KernelRow` using `user_id`.

File: src/ai/backend/manager/scheduler/dispatcher.py [hunk 5]
```diff
@@ -480,6 +481,10 @@ async def _check_predicates() -> List[Tuple[str, Union[Exception, PredicateResul
                                 "keypair_resource_limit",
                                 check_keypair_resource_limit(db_sess, sched_ctx, sess_ctx),
                             ),
+                            (
+                                "user_main_keypair_resource_limit",
```
Comment (achimnol) [inline src/ai/backend/manager/scheduler/dispatcher.py]: I think the name should be just "user_resource_limit".

File: src/ai/backend/manager/scheduler/predicates.py [hunk 6]
```diff
@@ -181,6 +182,47 @@ async def check_keypair_resource_limit(
     return PredicateResult(True)
 
 
+async def check_main_keypair_resource_limit(
```
Comment (achimnol) [inline src/ai/backend/manager/scheduler/predicates.py]: Same here.

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/manager/models/alembic/versions/d3f8c74bf148_user_main_keypair.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+            primary = pick_main_keypair(row.keypairs)
+            if primary is None:
+                # Create new keypair when the user has no keypair
+                kp_data = prepare_keypair(row.email, row.uuid, row.role)
+                db_session.execute(sa.insert(KeyPairRow).values(**kp_data))
+                user_id_pk_maps.append(
+                    {"user_id": row.uuid, "main_access_key": kp_data["access_key"]}
+                )
+            else:
+                user_id_pk_maps.append({"user_id": row.uuid, "main_access_key": primary.access_key})
```
Comment (achimnol) [inline src/ai/backend/manager/models/alembic/versions/d3f8c74bf148_user_main_keypair.py]: Just a very minor fix, but I think the variable `primary` should be `main_keypair` or `main_kp`. 😉

Comment (fregataa) [reply src/ai/backend/manager/models/alembic/versions/d3f8c74bf148_user_main_keypair.py]: oops, thank you. I forgot this part..

**Review by achimnol — APPROVED** [review]

**Pull request merged**